### ISSUE TITLE
UndefinedFunctionError instead of CompileError while calling unrequired Integer

### ISSUE THREAD

**Issue by koolhazcker** [elixir-lang/elixir#5438]
### Environment

* Elixir version (elixir -v): 
```
Erlang/OTP 19 [erts-8.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.3.4
```
* Operating system: macOS sierra

### Current behavior
```
iex(1)> Integer.is_odd(3)
** (UndefinedFunctionError) function Integer.is_odd/1 is undefined or private. Did you mean one of:

      * is_odd/1

    (elixir) Integer.is_odd(3)
iex(1)> require Integer
Integer
iex(2)> Integer.is_odd(3)
true
```
### Expected behavior
```
iex(1)> Integer.is_odd(3)
** (CompileError) iex:1: you must require Integer before invoking the macro Integer.is_odd/1
    (elixir) src/elixir_dispatch.erl:99: :elixir_dispatch.dispatch_require/6
iex(1)> require Integer
Integer
iex(2)> Integer.is_odd(3)
true
```

**Comment by josevalim** [comment]
We can only know if Integer.is_odd is a macro if the Integer module has been loaded. So we can't quite provide the expected behaviour you listed BUT we should improve the error message for this case.

**Comment by eksperimental** [comment]
error is correct.
Integer.is_odd/1 is a macro. and you need to require before using it.

**Comment by koolhazcker** [comment]
@josevalim maybe it is possible to make it more like `Logger` since it gives a more informative error message:
```
iex(3)> Logger.debug("2")
** (CompileError) iex:3: you must require Logger before invoking the macro Logger.debug/1
    (elixir) src/elixir_dispatch.erl:99: :elixir_dispatch.dispatch_require/6
```

**Comment by josevalim** [comment]
@koolhazcker that is only working because it happens that Logger has been loaded at this point. If Logger was not loaded, we wouldn't know.

**Comment by koolhazcker** [comment]
Ah, I understand now. Thanks for the explanation!

**Comment by eksperimental** [comment]
but since we are suggestion is_odd/1 we should say something like.

If you mean Integer.is_odd/1, you need to `require Integer` first.
or you might have mean one of  these:
- SomeModule.is_odd/1

Side note: it is weird that in the suggestions it doesn't mention the module name (it just says `is_odd/1` and not `Integer.is_odd/1`)

**Comment by josevalim** [comment]
@eksperimental  we don't need to mention the module name because the suggestions are specific to the given module, it would be redundant to do so. Other than that, that's the solution, yes. We need to see if what we are trying to call is a macro and, if so, act accordingly.

**Comment by lexmag** [comment]
It should already be solved via #5156, and available in master.

**Comment by josevalim** [comment]
Confirmed it has been fixed.
